workflows4s-web-ui/src/main/scala/workflows4s/web/ui/subs/InstancesManager.scala

```diff
@@ -297,45 +376,109 @@ final case class InstancesManager(
         code(instance.state.map(_.spaces2).getOrElse("No state available.")),
       ),
     )
+
+  private def generateMermaidFromProgress(progress: ProgressResponse): String = {
+    val builder = new StringBuilder()
+    builder.append("flowchart TD\n")
+    
+    // Start node
+    builder.append("  node_start@{ shape: circle, label: \"Start\"}\n")
+    
+    // Generate nodes for each step
+    val stepNodes = progress.steps.zipWithIndex.map { case (step, index) =>
+      val stepName = step.meta.name.getOrElse(s"Step ${index + 1}")
+      val nodeId = s"node_$index"
+      val isExecuted = step.result.exists(_.status == "Completed")
+      
+      val nodeShape = step.stepType match {
+        case "HandleSignal" => s"""$nodeId@{ shape: stadium, label: "fa:fa-envelope $stepName"}"""
+        case "Timer" => s"""$nodeId@{ shape: stadium, label: "fa:fa-clock $stepName"}"""
+        case "Pure" => s"""$nodeId["$stepName"]"""
+        case "RunIO" => s"""$nodeId["$stepName"]"""
+        case "Fork" => s"""$nodeId@{ shape: hex, label: "$stepName"}"""
+        case "Parallel" => s"""$nodeId@{ shape: fork, label: "$stepName"}"""
+        case "Retried" => s"""$nodeId["fa:fa-redo $stepName"]"""
+        case _ => s"""$nodeId["$stepName"]"""
+      }
+      
+      val nodeWithClass = if (isExecuted) nodeShape + ":::executed" else nodeShape
+      builder.append(s"  $nodeWithClass\n")
+      nodeId
+    }
+    
+    // End node
+    builder.append("  node_end@{ shape: circle, label: \"End\"}\n")
+    
+    // Generate connections
+    if (stepNodes.nonEmpty) {
+      builder.append(s"  node_start --> ${stepNodes.head}\n")
+      
+      for (i <- stepNodes.indices.init) {
+        builder.append(s"  ${stepNodes(i)} --> ${stepNodes(i + 1)}\n")
+      }
+      
+      builder.append(s"  ${stepNodes.last} --> node_end\n")
+    } else {
+      builder.append("  node_start --> node_end\n")
+    }
+    
+    // Add error handling connections if any steps have errors
+    val hasErrors = progress.steps.exists(_.meta.error.isDefined)
+    
+    if (hasErrors) {
+      builder.append("  node_error[\"Error Handler\"]\n")
+      progress.steps.zipWithIndex.foreach { case (step, index) =>
+        step.meta.error.foreach { errorName =>
+          builder.append(s"  node_$index -.->|\"fa:fa-bolt $errorName\"| node_error\n")
+        }
+      }
+      builder.append("  node_error --> node_end\n")
+    }
+    
+    // Add styling
+    builder.append("  classDef executed fill:#0e0\n")
+    
+    builder.toString()
+  }
 }
 
 object InstancesManager {
   def initial: InstancesManager =
     InstancesManager(
       instanceIdInput = "",
       state = State.Ready,
-      showJsonState = false,
-      progressState = ProgressState.Ready,
-      showMermaidViewer = false,
-      renderedMermaidSvg = None,
     )
 
   enum State {
```

Comment:
I think we seriously misunderstood each other. Why did you decide to build this kind of thing from scratch? 

What you had previously was perfectly fine. You could have simply include rendered mermaid diargram in `ProgressResponse` for example. Later we want to have there `WIOExecutionProgress` but right now already rendered mermaid would be just fine.